### PR TITLE
Test Component Rendering

### PR DESCRIPTION
**Pull request by na3d**
I have Navigation stashed, since testing it was dependent on what drew was working on, and working on actions and pulling in Navigation next.

### CODE REVIEW

File: src/components/Sidebar.js [hunk 1]
```diff
@@ -3,12 +3,11 @@ import { Link } from 'react-router';
 
 export default class Sidebar extends Component {
   render() {
-    const { username } = this.props;
```
Comment (SirCmpwn) [inline src/components/Sidebar.js]: Do you need to nuke the username in the props where Sidebar is invoked?

**Comment by SirCmpwn** [comment]
:+1:

Would like to see more comprehensive testing of these two components but would be reasonable to see that in a follow up PR.

File: src/components/Sidebar.js (hunk 1 above)
Comment (eatonphil) [inline src/components/Sidebar.js]: No https://github.com/linode/manager/blob/master/src/layouts/Layout.js#L13

Comment (na3d) [inline src/components/Sidebar.js]: username isn't being used, it is left over code from an earlier version.

Comment (SirCmpwn) [inline src/components/Sidebar.js]: kk

File: test/components/Dropdown.spec.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+import { expect } from 'chai';
+import Dropdown from '../../src/components/Dropdown';
+
+describe('Dropdown', () => {
+  it('renders dropdown component', () => {
+    const dropdown = renderIntoDocument(
+      <Dropdown elements={[
+        { _action: "ready", name: "Drew" },
+        { _action: "set", name: "Phil" },
+        { _action: "go", name: "Will" }
```
Comment (eatonphil) [inline test/components/Dropdown.spec.js]: Weren't we just talking about testing actions get called? Also, maybe mock up _action to something more similar to what it really is (a function for starters)?

Comment (eatonphil) [inline test/components/Dropdown.spec.js]: Also, _action is a random internal property used elsewhere that has nothing to do with Dropdown. The correct property name is `action`.

Comment (eatonphil) [inline test/components/Dropdown.spec.js]: https://github.com/linode/manager/blob/master/src/components/Dropdown.js#L11

Comment (SirCmpwn) [inline test/components/Dropdown.spec.js]: :+1: for declaring the actions correctly, but I wouldn't mind ensuring that they work in a seperate test/PR. This test is just asserting that the component was rendered correctly afaik

Comment (eatonphil) [inline test/components/Dropdown.spec.js]: Gotcha. Well yeah s/_action/action and set them to `() => {}` or something.

File: package.json [hunk 3]
```diff
@@ -84,6 +84,7 @@
     "react-redux": "^4.0.6",
     "react-router": "^1.0.3",
     "react-tabs": "^0.5.3",
+    "react-addons-test-utils": "^0.14.0",
```
Comment (eatonphil) [inline package.json]: I thought the point of bringing in phantomjs is that it does the rendering for us. Why are we having react do some rendering in these tests if we are going to use phantomjs? I may just be confused what the difference is.

Comment (SirCmpwn) [inline package.json]: I didn't mean rendering as in painting the DOM on the screen, I meant rendering as in turning `<Dropdown props... />` into a DOM tree and asserting that the DOM tree meets the dropdown spec (i.e. a list of the items you passed into the props).

Comment (SirCmpwn) [inline package.json]: i.e. the `render` method of react components

Comment (SirCmpwn) [inline package.json]: And the point of bringing in PhantomJS is that we can run tests for our browser app in a browser, and jsdom isn't necessary at that point.

**Comment by SirCmpwn** [comment]
Have you looked into [enzyme](http://airbnb.io/enzyme/)?

File: package.json (hunk 3 above)
Comment (eatonphil) [inline package.json]: Cool I think I get it.

**Comment by eatonphil** [comment]
I think this is ready to go now, right?

File: test/components/Navigation.spec.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+    expect(navigation.find('.nav-user').length).to.equal(1);
+  });
+
+  it('renders logo image', () => {
+    const navigation = mount(
+      <Navigation username={ 'peanut' } />
+    );
+    
+    expect(navigation.find('img').src).to.be.defined;
```
Comment (eatonphil) [inline test/components/Navigation.spec.js]: -1 what is the point?

File: karma.conf.js [hunk 5]
```diff
@@ -41,6 +41,12 @@ module.exports = function(config) {
       },
       sassLoader: {
         includePaths: [path.resolve(__dirname, "./node_modules/bootstrap/scss/")]
+      },
+      externals: {
+        'cheerio': 'window',
```
Comment (eatonphil) [inline karma.conf.js]: Why are we bringing in jquery?

File: test/components/Dropdown.spec.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    scryRenderedDOMComponentsWithClass
+} from 'react-addons-test-utils';
+import { expect } from 'chai';
+import Dropdown from '../../src/components/Dropdown';
+
+describe('Dropdown', () => {
+  it('renders dropdown component', () => {
+    const dropdown = renderIntoDocument(
+      <Dropdown elements={[
+        { action: function test(){}, name: "Drew" },
```
Comment (eatonphil) [inline test/components/Dropdown.spec.js]: s/function test(){}/() => {}/g

File: test/components/Dropdown.spec.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        { action: clickFirst, name: "Drew" },
+        { action: clickBodyItem, name: "Phil" },
+        { action: unclicked, name: "Will" }
+      ]} />
+    );
+
+    dropdown.find('.li-dropdown-first').first().simulate('click');
+    dropdown.find('.li-dropdown-body').find('.li-dropdown-item').first().simulate('click');
+
+    expect(clickFirst.calledOnce).to.equal(true);
```
Comment (eatonphil) [inline test/components/Dropdown.spec.js]: `s/to.equal(true)/to.be.equal/g`

File: test/components/Navigation.spec.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+import { expect } from 'chai';
+import Navigation from '../../src/components/Navigation';
+
+describe('Navigation', () => {
+  it('renders username in navigation component', () => {
+    const navigation = mount(
+      <Navigation username={ 'peanut' } />
+    );
+
+    expect(navigation.find('.nav-user').length).to.equal(1);
```
Comment (eatonphil) [inline test/components/Navigation.spec.js]: `expect(navigation.find('.nav-user').textContent).to.equal('peanut'))`

File: test/components/Sidebar.spec.js [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    );
+
+    const list = scryRenderedDOMComponentsWithTag(sidebar, 'li');
+
+    expect(list.length).to.equal(5);
+    expect(list[0].textContent).to.equal('Linodes');
+    expect(list[1].textContent).to.equal('NodeBalancers');
+    expect(list[2].textContent).to.equal('Longview');
+    expect(list[3].textContent).to.equal('DNS Manager');
+    expect(list[4].textContent).to.equal('Support');
```
Comment (eatonphil) [inline test/components/Sidebar.spec.js]: Again, what is the point here? This is the _static_ layout of the page and has nothing to do with any interactions or data. I don't understand why this should be tested.

File: test/components/Navigation.spec.js (hunk 4 above)
Comment (na3d) [inline test/components/Navigation.spec.js]: You're right, I was thinking to prove the logo image variable worked, but didn't remember to go back finish it.  I wanted to detect image not found, incase of a future refactor or other change.  Just went through to test, and it errors out in another way if the variable is wrong

File: karma.conf.js (hunk 5 above)
Comment (na3d) [inline karma.conf.js]: The externals is for handling a compatitbility issue.
http://airbnb.io/enzyme/docs/guides/webpack.html

File: test/components/Dropdown.spec.js (hunk 6 above)
Comment (eatonphil) [inline test/components/Dropdown.spec.js]: `s/action: function test(){}/action() {}/g`

Comment (eatonphil) [inline test/components/Dropdown.spec.js]: https://github.com/airbnb/javascript#es6-object-shorthand

File: karma.conf.js (hunk 5 above)
Comment (eatonphil) [inline karma.conf.js]: cool. looks good

File: test/components/Navigation.spec.js (hunk 8 above)
Comment (na3d) [inline test/components/Navigation.spec.js]: Surprisingly, that doesn't work.

File: test/components/Dropdown.spec.js (hunk 6 above)
Comment (SirCmpwn) [inline test/components/Dropdown.spec.js]: :+1:

File: test/components/Navigation.spec.js (hunk 8 above)
Comment (eatonphil) [inline test/components/Navigation.spec.js]: I was just guessing. Something like that will be a better test than passing in a specific username and seeing if some username exists.

File: test/components/Dropdown.spec.js (hunk 7 above)
Comment (SirCmpwn) [inline test/components/Dropdown.spec.js]: :-1:

Comment (eatonphil) [inline test/components/Dropdown.spec.js]: Wasn't thinking, just a sec

File: test/components/Navigation.spec.js (hunk 8 above)
Comment (SirCmpwn) [inline test/components/Navigation.spec.js]: :+1: to checking that the correct username is filled in

File: test/components/Dropdown.spec.js (hunk 7 above)
Comment (na3d) [inline test/components/Dropdown.spec.js]: Nice!

Comment (eatonphil) [inline test/components/Dropdown.spec.js]: `s/to.equal(true)/isTrue/g`

http://chaijs.com/api/assert/#method_istrue

File: test/components/Sidebar.spec.js (hunk 9 above)
Comment (SirCmpwn) [inline test/components/Sidebar.spec.js]: The navigation has a contract, or a spec if you prefer. It must:

1. List links to each of these pages
2. Show the logo
3. Show the username of the logged in user
4. Have a log out button?

And so on. We should test that the component fulfills its spec.

File: test/components/Dropdown.spec.js (hunk 7 above)
Comment (SirCmpwn) [inline test/components/Dropdown.spec.js]: :+1:

File: test/components/Sidebar.spec.js (hunk 9 above)
Comment (eatonphil) [inline test/components/Sidebar.spec.js]: I guess it's good to make sure that even things that will never change - do not actually change.

Comment (SirCmpwn) [inline test/components/Sidebar.spec.js]: Could also test that clicking each of these links dispatches a redux action to the router.

File: test/components/Dropdown.spec.js [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    const list = scryRenderedDOMComponentsWithClass(dropdown, 'li-dropdown-item');
+
+    expect(list.length).to.equal(3);
+    expect(list[0].textContent).to.equal('Drew');
+    expect(list[1].textContent).to.equal('Phil');
+    expect(list[2].textContent).to.equal('Will');
+  });
+
+  it('clickable dropdown', () => {
+    const clickFirst = sinon.spy();
```
Comment (eatonphil) [inline test/components/Dropdown.spec.js]: Good use of these here. 👍

File: test/components/Sidebar.spec.js (hunk 9 above)
Comment (na3d) [inline test/components/Sidebar.spec.js]: 👍   Yeah, and need to later detect that they don't 404.

File: test/components/Dropdown.spec.js (hunk 6 above)
Comment (SirCmpwn) [inline test/components/Dropdown.spec.js]: Clarification: :+1: to `() => {}`

Comment (eatonphil) [inline test/components/Dropdown.spec.js]: What's wrong with the function shorthand?

Comment (SirCmpwn) [inline test/components/Dropdown.spec.js]: I guess it just rubs me wrong. An object is key/value pairs, this syntax is weird and doesn't mesh with that well.

Comment (eatonphil) [inline test/components/Dropdown.spec.js]: ¯\_(ツ)_/¯

**Pull request merged**